wandb/sdk/internal/job_builder.py

```diff
@@ -0,0 +1,252 @@
+    _used_job: bool
+
+    def __init__(self, settings: SettingsStatic):
+        self._settings = settings
+        self._metadatafile_path = None
+        self._requirements_path = None
+        self._config = None
+        self._summary = None
+        self._logged_code_artifact = None
+        self._used_job = False
+
+    def set_config(self, config: Dict[str, Any]) -> None:
+        self._config = config
+
+    def set_summary(self, summary: Dict[str, Any]) -> None:
+        self._summary = summary
+
+    @property
+    def used_job(self) -> bool:
+        return self._used_job
+
+    def set_used_job(self, val: bool) -> None:
+        self._used_job = val
+
+    def _set_logged_code_artifact(
+        self, res: Optional[Dict], artifact: "ArtifactRecord"
+    ) -> None:
+        if artifact.type == "code" and res is not None:
+            self._logged_code_artifact = ArtifactInfoForJob(
+                {
+                    "id": res["id"],
+                    "name": artifact.name,
+                }
+            )
+
+    def _build_repo_job(
+        self, remote: str, commit: str, program_relpath: str, args: List[str]
+    ) -> Tuple[Artifact, GitSourceDict]:
+        source: GitSourceDict = {
+            "entrypoint": [
+                os.path.basename(sys.executable),
+                program_relpath,
+            ],
+            "args": args,
+            "git": {
+                "remote": remote,
+                "commit": commit,
+            },
+        }
+
+        name = make_artifact_name_safe(f"job-{remote}_{program_relpath}")
+
+        artifact = Artifact(name, JOB_ARTIFACT_TYPE)
+        if os.path.exists(os.path.join(self._settings.files_dir, DIFF_FNAME)):
+            artifact.add_file(
+                os.path.join(self._settings.files_dir, DIFF_FNAME),
+                name="diff.patch",
+            )
+        return artifact, source
+
+    def _build_artifact_job(
+        self, program_relpath: str, args: List[str]
+    ) -> Tuple[Artifact, ArtifactSourceDict]:
+        assert isinstance(self._logged_code_artifact, dict)
+        source: ArtifactSourceDict = {
+            "entrypoint": [
+                os.path.basename(sys.executable),
+                program_relpath,
+            ],
+            "args": args,
+            "artifact": f"wandb-artifact://_id/{self._logged_code_artifact['id']}",
+        }
+
+        name = f"job-{self._logged_code_artifact['name']}"
+
+        artifact = Artifact(name, JOB_ARTIFACT_TYPE)
+        return artifact, source
+
+    def _build_image_job(
+        self, image_name: str, args: List[str]
+    ) -> Tuple[Artifact, ImageSourceDict]:
+        name = make_artifact_name_safe(f"job-{image_name}")
+        artifact = Artifact(name, JOB_ARTIFACT_TYPE)
+        source: ImageSourceDict = {
+            "image": image_name,
+            "args": args,
+        }
+        return artifact, source
+
+    def build(self) -> Optional[Artifact]:
```

Comment:
maybe it makes sense to move some of the logic that extract the relevant data out of the build function and into the relevant build helper functions?